compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java

```diff
@@ -9504,17 +9501,35 @@ private STNode parseJoinClause(boolean isRhsExpr) {
     /**
      * Parse on clause.
      * <p>
-     * <code>on clause := on expression</code>
+     * <code>on clause := `on` expression `equals` expression</code>
      *
      * @return On clause node
      */
     private STNode parseOnClause(boolean isRhsExpr) {
+        STToken nextToken = peek();
+        if (isQueryClauseStartToken(nextToken.kind)) {
```

Comment:
Don't we need to recover here?
What if theres an additional token (say a `public` keyword) before the `on` keyword?

Reply:
shouldn't that extra token be recovered as invalid in parseOnKeyword() beneath?

Reply:
ah this is inverted. Sorry, my question should be inverted too :) 
What if theres an additional token before an actual query-clause starting token?
e.g:
```ballerina
from var person in personList
outer join Department dept in deptList
// missing on clause
someExtraToken select {
    fname : person.fname,
    lname : person.lname,
    dept : dept.name
};
```
will it keep the 'select' query as is?

Reply:
```
var a = from var person in personList
outer join Department dept in deptList
 MISSING[+]// missing on clause
someExtraToken  MISSING[on] MISSING[] MISSING[equals] MISSING[]select {
    fname : person.fname,
    lname : person.lname,
    dept : dept.name
};
```
Above is the recovery if that extra token is an identifier.  if it is `public`recovery is pretty bad yet it is a valid syntax. 
```
var a = from var person in personList
outer join Department dept in deptList
 MISSING[on] MISSING[<-]// missing on clause
 INVALID[public]  INVALID[select] {
    fname : person MISSING[,] INVALID[.]fname,
    lname : person MISSING[,] INVALID[.]lname,
    dept : dept MISSING[}].name
 MISSING[equals] MISSING[] MISSING[select] MISSING[] INVALID[}];
```
If we are going to handle when there's a  non-identifier token before the beginning of next clause in absence of join-on-condition it is like we are making join-on-condition optional in the error handler. As we don't allow that non-identifier to be part of join-on-condition. I think if there are extra tokens they should be allowed to be a part of join-on-condition recovery.
WDYT?

Reply:
We can do something like this in the error handler though. before parsing on-clause as next rule, we check if the next lookahead token is one of query starting tokens and if and only if, we skip join-on-condition path in the error handler. I feel that way recovery logic is consistence with parsing logic.

Reply:
hmm, I feel which ever option we take, there will be consequences.
We can go ahead with the current approach, and can revisit if it becomes an actual issue.